features/mbedtls/targets/TARGET_NUVOTON/TARGET_M480/sha/sha1_alt.c

```diff
@@ -14,53 +14,64 @@
  * limitations under the License.
  */
 
-#if !defined(MBEDTLS_CONFIG_FILE)
-#include "mbedtls/config.h"
-#else
-#include MBEDTLS_CONFIG_FILE
-#endif
+#include "mbedtls/sha1.h"
 
 #if defined(MBEDTLS_SHA1_C)
 #if defined(MBEDTLS_SHA1_ALT)
 
-#include "sha1_alt.h"
 #include "crypto-misc.h"
 #include "nu_bitutil.h"
 #include "string.h"
 
-void mbedtls_sha1_init(mbedtls_sha1_context *ctx)
+/* Choose SHA S/W or H/W context and initialize it
+ * 
+ * try_hw:
+ *   0: Initialize S/W context
+ *   1: Try acquiring SHA H/W resource first and initialize its H/W context if successful. If failed, initialize S/W context.
+ */
+static void mbedtls_sha1_init_internal(mbedtls_sha1_context *ctx, int try_hw)
 {
-    if (crypto_sha_acquire()) {
-        ctx->ishw = 1;
+    if (try_hw && crypto_sha_acquire()) {
+        ctx->active_ctx = &ctx->hw_ctx;
         mbedtls_sha1_hw_init(&ctx->hw_ctx);
-    }
-    else {
-        ctx->ishw = 0;
+    } else {
+        ctx->active_ctx = &ctx->sw_ctx;
         mbedtls_sha1_sw_init(&ctx->sw_ctx);
     }
 }
 
+void mbedtls_sha1_init(mbedtls_sha1_context *ctx)
+{
+    mbedtls_sha1_init_internal(ctx, 1);
+}
+
 void mbedtls_sha1_free(mbedtls_sha1_context *ctx)
 {
     if (ctx == NULL) {
         return;
     }
 
-    if (ctx->ishw) {
+    if (ctx->active_ctx == &ctx->hw_ctx) {
         mbedtls_sha1_hw_free(&ctx->hw_ctx);
         crypto_sha_release();
-    }
-    else {
+    } else if (ctx->active_ctx == &ctx->sw_ctx) {
         mbedtls_sha1_sw_free(&ctx->sw_ctx);
     }
+    ctx->active_ctx = NULL;
 }
 
 void mbedtls_sha1_clone(mbedtls_sha1_context *dst,
                         const mbedtls_sha1_context *src)
 {
```

Comment:
If `dst == src` and they are a H/W context, this function would currently free them. I suggest to return immediately in this case.

Reply:
@hanno-arm I fix it in https://github.com/ARMmbed/mbed-os/pull/4925/commits/f54aadcfe7cc8f9a6f9bf061cf29afa74f264c71.